java/ql/src/experimental/Security/CWE/CWE-522/InsecureBasicAuth.ql

```diff
@@ -0,0 +1,176 @@
+/**
+ * @name Insecure basic authentication
+ * @description Basic authentication only obfuscates username/password in Base64 encoding, which can be easily recognized and reversed. Transmission of sensitive information not over HTTPS is vulnerable to packet sniffing.
+ * @kind path-problem
+ * @id java/insecure-basic-auth
+ * @tags security
+ *       external/cwe-522
+ *       external/cwe-319
+ */
+
+import java
+import semmle.code.java.frameworks.Networking
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/**
+ * The Java class `org.apache.http.client.methods.HttpRequestBase`. Popular subclasses include `HttpGet`, `HttpPost`, and `HttpPut`.
+ * And the Java class `org.apache.http.message.BasicHttpRequest`.
+ */
+class ApacheHttpRequest extends RefType {
+  ApacheHttpRequest() {
+    this
+        .getASourceSupertype*()
+        .hasQualifiedName("org.apache.http.client.methods", "HttpRequestBase") or
+    this.getASourceSupertype*().hasQualifiedName("org.apache.http.message", "BasicHttpRequest")
+  }
+}
+
+/**
+ * Class of Java URL constructor.
+ */
+class URLConstructor extends ClassInstanceExpr {
+  URLConstructor() { this.getConstructor().getDeclaringType() instanceof TypeUrl }
+
+  Expr stringArg() {
+    // URLs constructed with any of the four string constructors below:
+    // URL(String spec), URL(String protocol, String host, int port, String file), URL(String protocol, String host, int port, String file, URLStreamHandler handler), URL(String protocol, String host, String file)
+    this.getConstructor().getParameter(0).getType() instanceof TypeString and
+    result = this.getArgument(0) // First argument contains the protocol part.
+  }
+}
+
+/**
+ * The type `java.net.URLConnection`.
+ */
+class TypeHttpUrlConnection extends RefType {
+  TypeHttpUrlConnection() {
+    this.getASourceSupertype*().hasQualifiedName("java.net", "HttpURLConnection")
+  }
+}
+
+/**
+ * String of HTTP URLs not in private domains.
+ */
+class HttpString extends StringLiteral {
+  HttpString() {
+    // Match URLs with the HTTP protocol and without private IP addresses to reduce false positives.
+    exists(string s | this.getRepresentedString() = s |
+      s = "http"
+      or
+      s.matches("http://%") and
+      not s.matches("%/localhost%") and
+      not s.matches("%/127.0.0.1%") and
+      not s.matches("%/10.%") and
+      not s.matches("%/172.16.%") and
+      not s.matches("%/192.168.%")
+    )
+  }
+}
+
+/**
+ * String concatenated with `HttpString`.
+ */
+predicate builtFromHttpStringConcat(Expr expr) {
+  expr instanceof HttpString
+  or
+  builtFromHttpStringConcat(expr.(AddExpr).getLeftOperand())
+  or
+  exists(Expr other | builtFromHttpStringConcat(other) |
+    exists(Variable var | var.getAnAssignedValue() = other and var.getAnAccess() = expr)
+  )
+  or
+  exists(Expr other | builtFromHttpStringConcat(other) |
```

Comment:
This looks quite complicated. I think you could achieve a similar effect if you let `builtFromHttpStringConcat` only find String values that look like `http:` URLs and add the `BasicRequestLine` and `URI` (and also `URL` ?) constructors as "additional taint steps" (by adding a method `isAdditionalTaintStep` on the dataflow configuration.)

Reply:
I've revamped the query as per the suggestion.